sql/analyzer/replace_pk_sort.go

```diff
@@ -123,3 +142,143 @@ func replacePkSortHelper(ctx *sql.Context, scope *plan.Scope, node sql.Node, sor
 	}
 	return newNode, transform.NewTree, nil
 }
+
+// replaceAgg converts aggregate functions to order by + limit 1 when possible
+func replaceAgg(ctx *sql.Context, a *Analyzer, node sql.Node, scope *plan.Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	return transform.Node(node, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		// project with groupby child
+		proj, ok := n.(*plan.Project)
+		if !ok {
+			return n, transform.SameTree, nil
+		}
+		gb, ok := proj.Child.(*plan.GroupBy)
+		if !ok {
+			return n, transform.SameTree, nil
+		}
+		// TODO: optimize when there are multiple aggregations; use LATERAL JOINS
+		if len(gb.SelectedExprs) != 1 || len(gb.GroupByExprs) != 0 {
+			return n, transform.SameTree, nil
+		}
+
+		var idxTbl sql.IndexAddressableTable
+		transform.Inspect(gb.Child, func(n sql.Node) bool {
+			switch t := gb.Child.(type) {
+			case *plan.IndexedTableAccess:
+				if tbl, ok := t.Table.(sql.IndexAddressableTable); ok {
+					idxTbl = tbl
+					return false
+				}
+			case *plan.ResolvedTable:
+				if tbl, ok := t.UnderlyingTable().(sql.IndexAddressableTable); ok {
+					idxTbl = tbl
+					return false
+				}
+			}
+			return true
+		})
+
+		// no indexes on table, unable to apply optimization
+		if idxTbl == nil {
+			return n, transform.SameTree, nil
+		}
+
+		// generate sort fields from aggregations
+		var sf sql.SortField
+		switch agg := gb.SelectedExprs[0].(type) {
+		case *aggregation.Max:
+			gf, ok := agg.UnaryExpression.Child.(*expression.GetField)
+			if !ok {
+				return n, transform.SameTree, nil
+			}
+			sf = sql.SortField{
+				Column: gf,
+				Order:  sql.Descending,
+			}
+		case *aggregation.Min:
+			gf, ok := agg.UnaryExpression.Child.(*expression.GetField)
+			if !ok {
+				return n, transform.SameTree, nil
+			}
+			sf = sql.SortField{
+				Column: gf,
+				Order:  sql.Ascending,
+			}
+		default:
+			return n, transform.SameTree, nil
+		}
+
+		// TODO: support secondary indexes
+		pkIdx, err := getPKIndex(ctx, idxTbl)
+		if err != nil {
+			return nil, transform.SameTree, err
+		}
+		if pkIdx == nil {
+			return n, transform.SameTree, nil
+		}
+		// since we're only supporting one aggregation, it must be on the first column of the primary key
+		if !strings.EqualFold(pkIdx.Expressions()[0], sf.Column.String()) {
```

Comment:
i think index expressions are table-prefixed, like `table.col`. Are the sort fields also always going to be prefixed the same way? Or this there a potential gap where one is table/col and the other is col?